src/v/cluster/cloud_storage_size_reducer.h

```diff
@@ -0,0 +1,196 @@
+#include "cluster/logger.h"
+#include "cluster/members_table.h"
+#include "cluster/topic_table.h"
+#include "cluster/topic_table_partition_generator.h"
+#include "config/node_config.h"
+#include "rpc/connection_cache.h"
+
+namespace cluster {
+
+class cloud_storage_size_reducer_exception : public std::runtime_error {
+public:
+    explicit cloud_storage_size_reducer_exception(const std::string& m)
+      : std::runtime_error(m) {}
+};
+
+/*
+ * This utility class performs a map-reduce operation accross the cluster
+ * in order to determine the total number of cloud storage bytes used by
+ * all partitions.
+ *
+ * It iterates across the current topic table in batche. For each batch,
+ * the following sequence of operations is performed:
+ * 1. Find the first live replica of each partition in the batch
+ * 2. Prepare `cloud_storage_usage` RPC requests for each node in the cluster.
+ * The request will contain the partitions being queried by shard.
+ * 3. Reduce the reponses and update the accumulator.
+ *
+ * One important thing to note is that this operation requires topic table
+ * stability (i.e. the topic table does not change in the meantime). This
+ * property is enfored by the `topic_table_partition_generator`, which will
+ * throw if the topic table has been updated during the iteration. If such
+ * a change does happen, it's treated as a retryable error.
+ */
+class cloud_storage_size_reducer {
+public:
+    static constexpr uint8_t default_retries_allowed = 3;
+
+    cloud_storage_size_reducer(
+      ss::sharded<topic_table>& topic_table,
+      ss::sharded<members_table>& members_table,
+      ss::sharded<rpc::connection_cache>& conn_cache,
+      size_t batch_size = topic_table_partition_generator::default_batch_size,
+      uint8_t retries_allowed = 3)
+      : _topic_table(std::ref(topic_table))
+      , _members_table(std::ref(members_table))
+      , _conn_cache{conn_cache}
+      , _batch_size{batch_size}
+      , _retries_allowed{retries_allowed} {}
+
+    ss::future<std::optional<uint64_t>> reduce() {
+        try {
+            co_return co_await do_reduce();
+        } catch (const topic_table_partition_generator_exception& e) {
+            vlog(
+              clusterlog.warn,
+              "Failed to generate total cloud storage usage across the cluster "
+              "(retries left: {}): {}",
+              _retries_allowed - _retries_attempted,
+              e.what());
+        } catch (const cloud_storage_size_reducer_exception& e) {
+            vlog(
+              clusterlog.warn,
+              "Failed to generate total cloud storage usage across the cluster "
+              "(retries left: {}): {}",
+              _retries_allowed - _retries_attempted,
+              e.what());
+        } catch (const std::exception& e) {
+            vlog(clusterlog.warn, "Unexpected exception thrown: {}", e.what());
+            throw;
+        }
+
+        if (++_retries_attempted <= _retries_allowed) {
```

Comment:
nit: `_retries_attempted` is never reset to 0 making me think this class is one-time-use as well. Could refactor into a function.

Reply:
This class is meant for one time use indeed. I've added a comment to note that. I can refactor it if you feel strongly about it.